<a href="https://colab.research.google.com/github/Athul-Kris/HEALTHCARE-CHECKER-SONAR/blob/main/ATHUL%20PYSCRIPTING.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## TASK 5 — Lambda, Filter, and Sorting Reports

This section focuses on creating analytics-style reports with concise functional programming.

### Task 5.1: Use a `lambda` + `sorted()` to display products sorted by price and stock level.

### Subtask:
1. Get a flat list of all products (reusing `get_all_products` from Task 2).
2. Sort this list by `price` using `lambda` and `sorted()`.
3. Sort the same list by `stock` level using `lambda` and `sorted()`.

**Reasoning**:
To sort products, I will first obtain a flat list of all products using the `get_all_products` function. Then, I will apply `sorted()` with a `lambda` function as the key to sort the products first by 'price' and then by 'stock' level. This demonstrates efficient in-memory sorting of complex data structures.

In [29]:
# Get all products using the previously defined function
all_products = get_all_products(inventory)

# Sort products by price
products_sorted_by_price = sorted(all_products, key=lambda product: product['price'])
print("\nProducts sorted by price:")
for product in products_sorted_by_price:
    print(f"- {product['name']}: Price {product['price']:.2f}")

# Sort products by stock level
products_sorted_by_stock = sorted(all_products, key=lambda product: product['stock'])
print("\nProducts sorted by stock level:")
for product in products_sorted_by_stock:
    print(f"- {product['name']}: Stock {product['stock']}")


Products sorted by price:
- T-Shirt: Price 25.99
- Jeans: Price 49.99
- Smartphone: Price 800.00
- Laptop: Price 1200.50

Products sorted by stock level:
- Laptop: Stock 14
- Smartphone: Stock 28
- Jeans: Stock 47
- T-Shirt: Stock 95


## Task 4.3: Use `json.dumps()` with `indent=4` to print a clean, readable report.

### Subtask:
Utilize the `json.dumps()` function with the `indent=4` parameter to print a pretty-printed JSON string representation of the `loaded_inventory` dictionary. This will serve as a clean, readable report of the current inventory data.

**Reasoning**:
To generate a readable JSON report, I will use `json.dumps()` on the `loaded_inventory` dictionary, specifying `indent=4` for pretty-printing. This will format the JSON output with a clear, hierarchical structure, making it easy to read.

In [28]:
import json

# Generate a pretty-printed JSON report of the loaded_inventory
pretty_json_report = json.dumps(loaded_inventory, indent=4)

print("\nPretty-printed JSON report of the loaded inventory:")
print(pretty_json_report)


Pretty-printed JSON report of the loaded inventory:
{
    "categories": {
        "Electronics": {
            "products": [
                {
                    "id": 102,
                    "name": "Smartphone",
                    "price": 800.0,
                    "stock": 28,
                    "suppliers": [
                        "MobileMasters"
                    ]
                },
                {
                    "id": 101,
                    "name": "Laptop",
                    "price": 1200.5,
                    "stock": 14,
                    "suppliers": [
                        "TechCorp",
                        "GadgetHub",
                        "NewSupplierCo"
                    ]
                }
            ]
        },
        "Clothing": {
            "products": [
                {
                    "id": 201,
                    "name": "T-Shirt",
                    "price": 25.99,
                    "stock": 95,
                    "su

## Task 4.2: Create a function `load_data(filename)`.

### Subtask:
Create a function `load_data(filename)` that takes a filename as input, reads the JSON data from that file, and loads it back into a Python dictionary.

**Reasoning**:
To implement `load_data`, I will define a function that takes a `filename` argument. Inside the function, I will open the specified file in read mode and use `json.load()` to parse the JSON content and return it as a Python dictionary. I will then test this function by loading `inventory.json`.

In [27]:
import json

def load_data(filename):
    with open(filename, 'r') as f:
        data = json.load(f)
    return data

# Test the function by loading inventory.json
loaded_inventory = load_data('inventory.json')

print("Inventory successfully loaded from 'inventory.json':")
display(loaded_inventory)

Inventory successfully loaded from 'inventory.json':


{'categories': {'Electronics': {'products': [{'id': 102,
     'name': 'Smartphone',
     'price': 800.0,
     'stock': 28,
     'suppliers': ['MobileMasters']},
    {'id': 101,
     'name': 'Laptop',
     'price': 1200.5,
     'stock': 14,
     'suppliers': ['TechCorp', 'GadgetHub', 'NewSupplierCo']}]},
  'Clothing': {'products': [{'id': 201,
     'name': 'T-Shirt',
     'price': 25.99,
     'stock': 95,
     'suppliers': ['FashionInc', 'StyleWear']},
    {'id': 202,
     'name': 'Jeans',
     'price': 49.99,
     'stock': 47,
     'suppliers': ['DenimCo']}]}}}

## TASK 4 — JSON and Data Serialization

This section focuses on saving and loading inventory data using JSON, as well as performing advanced data transformations.

### Task 4.1: Export the entire inventory dictionary to a file `inventory.json`.

**Reasoning**:
To export the `inventory` data to a JSON file, I will import the `json` module. Then, I will open a file named `inventory.json` in write mode and use `json.dump()` to serialize the `inventory` dictionary into JSON format and write it to the file. This creates a persistent record of the current inventory state.

In [26]:
import json

# Export the inventory dictionary to inventory.json
with open('inventory.json', 'w') as f:
    json.dump(inventory, f, indent=4)

print("Inventory successfully exported to 'inventory.json'.")

Inventory successfully exported to 'inventory.json'.


## TASK 1 — Data Setup & Nested Structures

First, let's define the initial `inventory` data structure as provided.

In [1]:
inventory = {
    "categories": {
        "Electronics": {
            "products": [
                {"id": 101, "name": "Laptop", "price": 1200.50, "stock": 15, "suppliers": ["TechCorp", "GadgetHub"]},
                {"id": 102, "name": "Smartphone", "price": 800.00, "stock": 30, "suppliers": ["MobileMasters"]}
            ]
        },
        "Clothing": {
            "products": [
                {"id": 201, "name": "T-Shirt", "price": 25.99, "stock": 100, "suppliers": ["FashionInc", "StyleWear"]},
                {"id": 202, "name": "Jeans", "price": 49.99, "stock": 50, "suppliers": ["DenimCo"]}
            ]
        }
    }
}

print("Initial Inventory:")
display(inventory)

Initial Inventory:


{'categories': {'Electronics': {'products': [{'id': 101,
     'name': 'Laptop',
     'price': 1200.5,
     'stock': 15,
     'suppliers': ['TechCorp', 'GadgetHub']},
    {'id': 102,
     'name': 'Smartphone',
     'price': 800.0,
     'stock': 30,
     'suppliers': ['MobileMasters']}]},
  'Clothing': {'products': [{'id': 201,
     'name': 'T-Shirt',
     'price': 25.99,
     'stock': 100,
     'suppliers': ['FashionInc', 'StyleWear']},
    {'id': 202,
     'name': 'Jeans',
     'price': 49.99,
     'stock': 50,
     'suppliers': ['DenimCo']}]}}}

### Task 1.1: Print all product names across categories using nested loops.

In [2]:
print("\nProduct Names:")
for category_name, category_data in inventory["categories"].items():
    for product in category_data["products"]:
        print(f"- {product['name']} ({category_name})")


Product Names:
- Laptop (Electronics)
- Smartphone (Electronics)
- T-Shirt (Clothing)
- Jeans (Clothing)


### Task 1.2: Use list comprehension to extract a flat list of all supplier names.

In [3]:
all_suppliers = [
    supplier
    for category_data in inventory["categories"].values()
    for product in category_data["products"]
    for supplier in product["suppliers"]
]

print("\nAll Supplier Names:")
print(all_suppliers)


All Supplier Names:
['TechCorp', 'GadgetHub', 'MobileMasters', 'FashionInc', 'StyleWear', 'DenimCo']


### Task 1.3: Add a new product record to the Electronics category.

In [4]:
new_product = {"id": 103, "name": "Smartwatch", "price": 250.00, "stock": 20, "suppliers": ["WearablesInc"]}
inventory["categories"]["Electronics"]["products"].append(new_product)

print("\nInventory after adding Smartwatch:")
display(inventory)


Inventory after adding Smartwatch:


{'categories': {'Electronics': {'products': [{'id': 101,
     'name': 'Laptop',
     'price': 1200.5,
     'stock': 15,
     'suppliers': ['TechCorp', 'GadgetHub']},
    {'id': 102,
     'name': 'Smartphone',
     'price': 800.0,
     'stock': 30,
     'suppliers': ['MobileMasters']},
    {'id': 103,
     'name': 'Smartwatch',
     'price': 250.0,
     'stock': 20,
     'suppliers': ['WearablesInc']}]},
  'Clothing': {'products': [{'id': 201,
     'name': 'T-Shirt',
     'price': 25.99,
     'stock': 100,
     'suppliers': ['FashionInc', 'StyleWear']},
    {'id': 202,
     'name': 'Jeans',
     'price': 49.99,
     'stock': 50,
     'suppliers': ['DenimCo']}]}}}

### Task 1.4: Decrease every product's stock by 5% (simulate sales) using nested iteration.

In [5]:
for category_name, category_data in inventory["categories"].items():
    for product in category_data["products"]:
        product['stock'] = int(product['stock'] * 0.95) # Decrease by 5% and cast to int

print("\nInventory after decreasing stock by 5%:")
display(inventory)


Inventory after decreasing stock by 5%:


{'categories': {'Electronics': {'products': [{'id': 101,
     'name': 'Laptop',
     'price': 1200.5,
     'stock': 14,
     'suppliers': ['TechCorp', 'GadgetHub']},
    {'id': 102,
     'name': 'Smartphone',
     'price': 800.0,
     'stock': 28,
     'suppliers': ['MobileMasters']},
    {'id': 103,
     'name': 'Smartwatch',
     'price': 250.0,
     'stock': 19,
     'suppliers': ['WearablesInc']}]},
  'Clothing': {'products': [{'id': 201,
     'name': 'T-Shirt',
     'price': 25.99,
     'stock': 95,
     'suppliers': ['FashionInc', 'StyleWear']},
    {'id': 202,
     'name': 'Jeans',
     'price': 49.99,
     'stock': 47,
     'suppliers': ['DenimCo']}]}}}

### Task 1.5: Use a lambda to sort products (in each category) by price.

In [6]:
for category_name, category_data in inventory["categories"].items():
    category_data["products"].sort(key=lambda product: product['price'])

print("\nInventory after sorting products by price in each category:")
display(inventory)


Inventory after sorting products by price in each category:


{'categories': {'Electronics': {'products': [{'id': 103,
     'name': 'Smartwatch',
     'price': 250.0,
     'stock': 19,
     'suppliers': ['WearablesInc']},
    {'id': 102,
     'name': 'Smartphone',
     'price': 800.0,
     'stock': 28,
     'suppliers': ['MobileMasters']},
    {'id': 101,
     'name': 'Laptop',
     'price': 1200.5,
     'stock': 14,
     'suppliers': ['TechCorp', 'GadgetHub']}]},
  'Clothing': {'products': [{'id': 201,
     'name': 'T-Shirt',
     'price': 25.99,
     'stock': 95,
     'suppliers': ['FashionInc', 'StyleWear']},
    {'id': 202,
     'name': 'Jeans',
     'price': 49.99,
     'stock': 47,
     'suppliers': ['DenimCo']}]}}}

# Task
Perform advanced data manipulation on the provided `inventory` dictionary: find all products supplied by 'TechCorp', calculate the total value of stock for each category, identify products with stock levels below 20, create a new nested dictionary grouping products by supplier, and remove 'Smartwatch' from the 'Electronics' category. Finally, summarize the manipulations performed and the insights gained.

## Task 2.1: Find all products supplied by 'TechCorp'.

### Subtask:
Iterate through the inventory and identify products that list 'TechCorp' as one of their suppliers. The result should be a list of product names.


**Reasoning**:
To identify products supplied by 'TechCorp', I will iterate through the nested `inventory` structure, check each product's suppliers, and collect the names of products where 'TechCorp' is listed.



In [7]:
techcorp_products = []

for category_name, category_data in inventory["categories"].items():
    for product in category_data["products"]:
        if "TechCorp" in product["suppliers"]:
            techcorp_products.append(product["name"])

print("\nProducts supplied by TechCorp:")
print(techcorp_products)


Products supplied by TechCorp:
['Laptop']


## Task 2.2: Calculate the total value of stock for each category.

### Subtask:
For each category, sum the (price * stock) for all products to get the total stock value. The output should be a dictionary with category names as keys and their total stock values as values.


**Reasoning**:
I need to calculate the total stock value for each category by iterating through the inventory, multiplying price by stock for each product, and summing these values per category.



In [8]:
category_stock_value = {}

for category_name, category_data in inventory["categories"].items():
    total_value = 0
    for product in category_data["products"]:
        total_value += product['price'] * product['stock']
    category_stock_value[category_name] = total_value

print("\nTotal Stock Value per Category:")
print(category_stock_value)


Total Stock Value per Category:
{'Electronics': 43957.0, 'Clothing': 4818.58}


## Task 2.3: Identify products with stock levels below 20.

### Subtask:
Find all products across all categories that have a 'stock' value less than 20. The output should be a list of product names.

### Reasoning:
To identify products with low stock, I will iterate through all categories and their products, checking each product's 'stock' level against the threshold of 20. Products meeting this condition will have their names added to a list.

**Reasoning**:
To identify products with low stock, I will iterate through the `inventory` dictionary, checking each product's 'stock' level against the threshold of 20 and collect the names of products that meet this criterion.



In [9]:
low_stock_products = []

for category_data in inventory["categories"].values():
    for product in category_data["products"]:
        if product['stock'] < 20:
            low_stock_products.append(product['name'])

print("\nProducts with stock less than 20:")
print(low_stock_products)


Products with stock less than 20:
['Smartwatch', 'Laptop']


## Task 2.4: Create a new nested dictionary grouping products by supplier.

### Subtask:
Create a dictionary where keys are supplier names and values are lists of product names that each supplier provides. Each product name should only appear once per supplier.

### Reasoning:
To group products by supplier, I will iterate through all categories and products. For each product, I will iterate through its suppliers and add the product's name to a list associated with that supplier in a new dictionary. I will use a set to ensure unique product names per supplier initially, then convert to a list for the final output.

## Task 2.4: Create a new nested dictionary grouping products by supplier.

### Subtask:
Create a dictionary where keys are supplier names and values are lists of product names that each supplier provides. Each product name should only appear once per supplier.

### Reasoning:
To group products by supplier, I will iterate through all categories and products. For each product, I will iterate through its suppliers and add the product's name to a list associated with that supplier in a new dictionary. I will use a set to ensure unique product names per supplier initially, then convert to a list for the final output.

**Reasoning**:
To group products by supplier, I will iterate through all categories and products. For each product, I will iterate through its suppliers and add the product's name to a set associated with that supplier in a temporary dictionary to ensure uniqueness. Finally, I will convert these sets to lists for the final output.



In [10]:
products_by_supplier = {}

for category_data in inventory["categories"].values():
    for product in category_data["products"]:
        product_name = product['name']
        for supplier in product['suppliers']:
            if supplier not in products_by_supplier:
                products_by_supplier[supplier] = []
            if product_name not in products_by_supplier[supplier]: # Ensure uniqueness
                products_by_supplier[supplier].append(product_name)

print("\nProducts Grouped by Supplier:")
print(products_by_supplier)


Products Grouped by Supplier:
{'WearablesInc': ['Smartwatch'], 'MobileMasters': ['Smartphone'], 'TechCorp': ['Laptop'], 'GadgetHub': ['Laptop'], 'FashionInc': ['T-Shirt'], 'StyleWear': ['T-Shirt'], 'DenimCo': ['Jeans']}


## Task 2.5: Remove 'Smartwatch' from the 'Electronics' category.

### Subtask:
Locate and remove the product 'Smartwatch' from the `products` list within the 'Electronics' category in the `inventory` dictionary.

### Reasoning:
To remove 'Smartwatch', I will access the 'Electronics' category and iterate through its 'products' list to find the product with the name 'Smartwatch'. Once found, I will remove it from the list.

**Reasoning**:
To remove the 'Smartwatch' product, I will iterate through the products in the 'Electronics' category and use a list comprehension or a loop with a conditional check to create a new list excluding the product with the name 'Smartwatch'.



In [11]:
electronics_products = inventory["categories"]["Electronics"]["products"]
inventory["categories"]["Electronics"]["products"] = [
    product for product in electronics_products if product["name"] != "Smartwatch"
]

print("\nInventory after removing Smartwatch:")
display(inventory)


Inventory after removing Smartwatch:


{'categories': {'Electronics': {'products': [{'id': 102,
     'name': 'Smartphone',
     'price': 800.0,
     'stock': 28,
     'suppliers': ['MobileMasters']},
    {'id': 101,
     'name': 'Laptop',
     'price': 1200.5,
     'stock': 14,
     'suppliers': ['TechCorp', 'GadgetHub']}]},
  'Clothing': {'products': [{'id': 201,
     'name': 'T-Shirt',
     'price': 25.99,
     'stock': 95,
     'suppliers': ['FashionInc', 'StyleWear']},
    {'id': 202,
     'name': 'Jeans',
     'price': 49.99,
     'stock': 47,
     'suppliers': ['DenimCo']}]}}}

## Task Summary and Insights Gained

This task involved performing several advanced data manipulation operations on a nested `inventory` dictionary. The following operations were successfully carried out:

1.  **Identified Products by Supplier**: We found all products supplied by 'TechCorp', resulting in `['Laptop']`.

2.  **Calculated Total Stock Value per Category**: We computed the total monetary value of stock for each category, yielding `{'Electronics': 43957.0, 'Clothing': 4818.58}`. This gives a quick overview of the value tied up in each product category.

3.  **Identified Low Stock Products**: Products with a stock level below 20 were identified, which were `['Smartwatch', 'Laptop']`. This is crucial for inventory management and reordering processes.

4.  **Grouped Products by Supplier**: A new dictionary was created to group products by their suppliers, showing which supplier provides which products. For example, `{'WearablesInc': ['Smartwatch'], 'MobileMasters': ['Smartphone'], 'TechCorp': ['Laptop'], 'GadgetHub': ['Laptop'], 'FashionInc': ['T-Shirt'], 'StyleWear': ['T-Shirt'], 'DenimCo': ['Jeans']}`. This can be useful for supplier relationship management and identifying single points of failure.

5.  **Removed a Product**: The 'Smartwatch' product was successfully removed from the 'Electronics' category.

### Key Insights:
*   **Inventory Health**: The low stock product identification (`'Smartwatch'`, `'Laptop'`) immediately flags items that might need attention to prevent stockouts.
*   **Value Distribution**: The category stock value shows that 'Electronics' holds significantly more value than 'Clothing', which could inform financial planning and investment decisions.
*   **Supplier Dependence**: The grouping by supplier helps in understanding supplier relationships and potentially negotiating better terms or diversifying suppliers if too many critical products come from a single source.
*   **Dynamic Data Management**: The exercise demonstrated the flexibility and power of Python's dictionary and list operations for managing complex, nested data structures efficiently.

## Final Task

### Subtask:
Summarize the advanced data manipulation tasks performed and the insights gained from each task.


## Summary:

### Q&A
The task involved summarizing the advanced data manipulation tasks performed and the insights gained from each task.

### Data Analysis Key Findings
*   **Products by Supplier**: Only 'Laptop' was found to be supplied by 'TechCorp'.
*   **Category Stock Value**: The total stock value for 'Electronics' was \$43957.0, while for 'Clothing' it was \$4818.58, indicating a much higher monetary value in the Electronics category.
*   **Low Stock Products**: 'Smartwatch' and 'Laptop' were identified as products with stock levels below 20, signaling potential items for reordering.
*   **Products Grouped by Supplier**: A new dictionary was successfully created, showing which suppliers provide which products (e.g., 'WearablesInc' supplies 'Smartwatch', 'TechCorp' supplies 'Laptop', 'FashionInc' supplies 'T-Shirt').
*   **Product Removal**: The product 'Smartwatch' was successfully removed from the 'Electronics' category in the inventory.

### Insights or Next Steps
*   The identification of low-stock products like 'Smartwatch' and 'Laptop' directly flags items requiring immediate attention to prevent stockouts and ensure continuous availability.
*   The significant difference in total stock value between 'Electronics' (\$43957.0) and 'Clothing' (\$4818.58) highlights where the business's capital is primarily invested, which can inform financial planning and inventory investment strategies.


# Task
Implement the following Python functions for functional programming on the `inventory` data structure: `get_all_products(data)` to return a flat list of all product dictionaries; `get_total_stock(cat_name)` to calculate the total stock for a given category; `add_supplier(prod_id, supplier_name)` to add a supplier to a product; `filter_high_price(threshold)` using `filter()` and `lambda` to find products above a price threshold; and `get_avg_price()` using `map()` and `functools.reduce()` to compute the inventory-wide average product price. Finally, summarize the purpose and utility of these functions.

## Define get_all_products(data)

### Subtask:
Create a function `get_all_products(data)` that takes the inventory data as input and returns a flat list containing all product dictionaries from across all categories.


**Reasoning**:
I need to define the `get_all_products` function as described, iterating through the nested dictionary to collect all product dictionaries into a single list.



In [12]:
def get_all_products(data):
    all_products = []
    for category_data in data["categories"].values():
        all_products.extend(category_data["products"])
    return all_products

# Test the function with the existing inventory data
all_inventory_products = get_all_products(inventory)

print("All products in a flat list:")
display(all_inventory_products)


All products in a flat list:


[{'id': 102,
  'name': 'Smartphone',
  'price': 800.0,
  'stock': 28,
  'suppliers': ['MobileMasters']},
 {'id': 101,
  'name': 'Laptop',
  'price': 1200.5,
  'stock': 14,
  'suppliers': ['TechCorp', 'GadgetHub']},
 {'id': 201,
  'name': 'T-Shirt',
  'price': 25.99,
  'stock': 95,
  'suppliers': ['FashionInc', 'StyleWear']},
 {'id': 202,
  'name': 'Jeans',
  'price': 49.99,
  'stock': 47,
  'suppliers': ['DenimCo']}]

## Define get_total_stock(cat_name)

### Subtask:
Create a function `get_total_stock(cat_name)` that takes a category name as input and returns the total stock count for all products within that specific category.


### Reasoning:
To calculate the total stock for a specific category, I will define a function `get_total_stock` that takes the inventory data and the category name as input. Inside the function, I will iterate through the products of the specified category, summing up their 'stock' values, and then return the accumulated total.

**Reasoning**:
Now I will implement the `get_total_stock` function as described, including its definition and a test call with 'Electronics' category to demonstrate its functionality.



In [13]:
def get_total_stock(data, cat_name):
    total_stock = 0
    if cat_name in data["categories"]:
        for product in data["categories"][cat_name]["products"]:
            total_stock += product['stock']
    return total_stock

# Test the function for 'Electronics' category
electronics_total_stock = get_total_stock(inventory, 'Electronics')
print(f"Total stock for Electronics category: {electronics_total_stock}")

# Test the function for 'Clothing' category
clothing_total_stock = get_total_stock(inventory, 'Clothing')
print(f"Total stock for Clothing category: {clothing_total_stock}")

Total stock for Electronics category: 42
Total stock for Clothing category: 142


## Define add_supplier(prod_id, supplier_name)

### Subtask:
Create a function `add_supplier(prod_id, supplier_name)` that takes a product ID and a supplier name as input. This function should locate the product by its ID and add the new supplier name to its `suppliers` list. Ensure that the supplier is not added if it already exists for that product.

### Reasoning:
To implement `add_supplier`, I will iterate through all products in the `inventory` to find the one matching the given `prod_id`. Once found, I will check if the `supplier_name` is already in its `suppliers` list; if not, I will append the new supplier. This ensures no duplicate suppliers are added.

**Reasoning**:
Now I will implement the `add_supplier` function as described, and test it by adding a new supplier to an existing product.



In [14]:
def add_supplier(data, prod_id, supplier_name):
    for category_data in data["categories"].values():
        for product in category_data["products"]:
            if product["id"] == prod_id:
                if supplier_name not in product["suppliers"]:
                    product["suppliers"].append(supplier_name)
                    print(f"Added '{supplier_name}' to product '{product["name"]}' (ID: {prod_id}).")
                else:
                    print(f"Supplier '{supplier_name}' already exists for product '{product["name"]}' (ID: {prod_id}).")
                return True
    print(f"Product with ID {prod_id} not found.")
    return False

# Test adding a new supplier to 'Laptop' (ID 101)
add_supplier(inventory, 101, "NewSupplierCo")

# Test adding an existing supplier to 'Smartphone' (ID 102)
add_supplier(inventory, 102, "MobileMasters")

# Display updated inventory to verify
print("\nInventory after adding suppliers:")
display(inventory)

Added 'NewSupplierCo' to product 'Laptop' (ID: 101).
Supplier 'MobileMasters' already exists for product 'Smartphone' (ID: 102).

Inventory after adding suppliers:


{'categories': {'Electronics': {'products': [{'id': 102,
     'name': 'Smartphone',
     'price': 800.0,
     'stock': 28,
     'suppliers': ['MobileMasters']},
    {'id': 101,
     'name': 'Laptop',
     'price': 1200.5,
     'stock': 14,
     'suppliers': ['TechCorp', 'GadgetHub', 'NewSupplierCo']}]},
  'Clothing': {'products': [{'id': 201,
     'name': 'T-Shirt',
     'price': 25.99,
     'stock': 95,
     'suppliers': ['FashionInc', 'StyleWear']},
    {'id': 202,
     'name': 'Jeans',
     'price': 49.99,
     'stock': 47,
     'suppliers': ['DenimCo']}]}}}

## Define filter_high_price(threshold)

### Subtask:
Create a function `filter_high_price(threshold)` that takes a price `threshold` as input. This function should use `filter()` and a `lambda` function to return a list of product names whose prices are greater than the given `threshold`.

### Reasoning:
To implement `filter_high_price`, I will first flatten the inventory into a list of all products. Then, I will use Python's built-in `filter()` function along with a `lambda` expression to select products whose price exceeds the given `threshold`. Finally, I will extract and return the names of these filtered products.

**Reasoning**:
Now I will implement the `filter_high_price` function using `filter()` and a `lambda` function as described, and test it with an example price threshold.



In [15]:
def filter_high_price(data, threshold):
    all_products = get_all_products(data)
    high_price_products = list(filter(lambda product: product['price'] > threshold, all_products))
    return [product['name'] for product in high_price_products]

# Test the function with a threshold of 500
high_price_product_names = filter_high_price(inventory, 500)
print(f"\nProducts with price greater than 500: {high_price_product_names}")

# Test the function with a threshold of 100
high_price_product_names_100 = filter_high_price(inventory, 100)
print(f"Products with price greater than 100: {high_price_product_names_100}")


Products with price greater than 500: ['Smartphone', 'Laptop']
Products with price greater than 100: ['Smartphone', 'Laptop']


## Define get_avg_price()

### Subtask:
Create a function `get_avg_price()` that computes the inventory-wide average product price. This function should use `map()` to extract prices and `functools.reduce()` to sum them, then divide by the total number of products. Make sure to import `functools`.


### Reasoning:
To calculate the inventory-wide average product price, I will first get a flat list of all products. Then, I will use `map()` to extract the 'price' of each product. Finally, I will use `functools.reduce()` to sum these prices and divide by the total number of products to get the average. I will need to import the `functools` module for `reduce`.

**Reasoning**:
Now I will implement the `get_avg_price` function using `map()` and `functools.reduce()` as described, and test it to compute the average product price.



In [16]:
import functools

def get_avg_price(data):
    all_products = get_all_products(data)
    if not all_products:
        return 0.0

    # Extract prices using map
    prices = list(map(lambda product: product['price'], all_products))

    # Sum prices using functools.reduce
    total_price = functools.reduce(lambda x, y: x + y, prices)

    # Calculate average
    average_price = total_price / len(all_products)
    return average_price

# Test the function
avg_price = get_avg_price(inventory)
print(f"\nInventory-wide average product price: {avg_price:.2f}")


Inventory-wide average product price: 519.12


## Summary of Functional Programming Functions

This section aimed to demonstrate functional programming concepts using the `inventory` data structure. The following functions were created:

1.  **`get_all_products(data)`**: This function returns a flat list of all product dictionaries across all categories. Its utility lies in providing a simplified view of all products, which is often a necessary first step for global operations or when an aggregate view is needed without regard to category structure.

2.  **`get_total_stock(data, cat_name)`**: This function calculates the total stock count for a specified category. It's useful for quick category-level inventory checks, allowing for focused stock management and reporting for particular product groups.

3.  **`add_supplier(data, prod_id, supplier_name)`**: This function allows for dynamically updating product information by adding a new supplier to a product, ensuring no duplicates are added. This highlights the ability to modify nested data structures safely and efficiently, which is crucial for data maintenance.

4.  **`filter_high_price(data, threshold)`**: Utilizing `filter()` and a `lambda` function, this function extracts product names that exceed a given price threshold. It's a powerful way to quickly identify premium products, perform market analysis, or flag items for specific pricing strategies.

5.  **`get_avg_price(data)`**: This function computes the inventory-wide average product price using `map()` to extract prices and `functools.reduce()` to sum them. This demonstrates efficient aggregate calculations over large datasets, providing a key performance indicator for the overall product catalog.

### Overall Utility:
These functions collectively illustrate how functional programming paradigms (like using `map`, `filter`, `reduce`, and `lambda` functions) can be applied to complex nested data structures in Python. They promote code reusability, readability, and efficiency for common data manipulation tasks, making the `inventory` management more modular and easier to maintain.

## Final Task

### Subtask:
Summarize the functions defined for the Functional Programming Layer and their utility.


## Summary:

### Q&A
The task asked to summarize the functions defined for the Functional Programming Layer and their utility.

1.  **`get_all_products(data)`**: This function provides a flattened list of all product dictionaries from the inventory, simplifying operations that require an aggregate view of all products regardless of their category.
2.  **`get_total_stock(data, cat_name)`**: This utility function calculates the sum of stock for all products within a specified category, enabling focused inventory checks and reporting.
3.  **`add_supplier(data, prod_id, supplier_name)`**: This function allows for the dynamic update of product information by adding new suppliers to a product's list, while preventing duplicate entries. It's crucial for maintaining accurate product data.
4.  **`filter_high_price(data, threshold)`**: Using `filter()` and a `lambda` function, this function identifies and lists product names whose prices exceed a given threshold. It is useful for market analysis, identifying premium items, or specific pricing strategies.
5.  **`get_avg_price(data)`**: This function computes the average price of all products across the entire inventory using `map()` and `functools.reduce()`, providing a key performance indicator for the overall product catalog.

### Data Analysis Key Findings

*   The `get_all_products` function successfully extracted all products, creating a flat list of 4 product dictionaries.
*   The `get_total_stock` function correctly calculated the total stock:
    *   'Electronics' category: 42 units.
    *   'Clothing' category: 142 units.
*   The `add_supplier` function demonstrated successful modification and duplicate prevention:
    *   'NewSupplierCo' was successfully added to 'Laptop' (ID 101).
    *   Adding 'MobileMasters' to 'Smartphone' (ID 102) was skipped as the supplier already existed for that product.
*   The `filter_high_price` function identified products based on price thresholds:
    *   Products with a price greater than \$500: \['Smartphone', 'Laptop'].
    *   Products with a price greater than \$100: \['Smartphone', 'Laptop'].
*   The `get_avg_price` function accurately computed the inventory-wide average product price as \$519.12.

### Insights or Next Steps

*   The implemented functions effectively leverage functional programming paradigms like `map`, `filter`, `reduce`, and `lambda` to perform common data manipulation and aggregation tasks on complex nested data structures, promoting code reusability and clarity.
*   These functions provide a robust foundation for building more advanced inventory management features, such as dynamic reporting, automated stock alerts, or supplier-specific analyses.


# Task
Define a Python class `ProductIterator` that implements the iterator protocol (`__iter__` and `__next__`) to iterate through and return the names of all products in the `inventory` one by one.

## Task 3.1: Create a custom iterator class ProductIterator.

### Subtask:
Define a Python class `ProductIterator` that implements the iterator protocol (`__iter__` and `__next__`) to iterate through and return the names of all products in the `inventory` one by one.


### Reasoning:
To create the `ProductIterator` class, I will first define its `__init__` method to flatten the nested `inventory` dictionary into a list of all product names and initialize a pointer for iteration. Then, I will implement `__iter__` to return the instance itself, making it an iterator. Finally, the `__next__` method will sequentially return product names, raising `StopIteration` when all products have been yielded, adhering to Python's iterator protocol.

**Reasoning**:
Now I will implement the `ProductIterator` class as described, including its `__init__`, `__iter__`, and `__next__` methods, and then test it by iterating through an instance to print all product names.



In [17]:
class ProductIterator:
    def __init__(self, data):
        self._product_names = []
        for category_data in data["categories"].values():
            for product in category_data["products"]:
                self._product_names.append(product["name"])
        self._index = 0

    def __iter__(self):
        return self

    def __next__(self):
        if self._index < len(self._product_names):
            product_name = self._product_names[self._index]
            self._index += 1
            return product_name
        else:
            raise StopIteration

# Create an instance of ProductIterator
product_iterator = ProductIterator(inventory)

print("\nIterating through product names using ProductIterator:")
for name in product_iterator:
    print(f"- {name}")


Iterating through product names using ProductIterator:
- Smartphone
- Laptop
- T-Shirt
- Jeans


## Task 3.2: Build a generator function supplier_generator().

### Subtask:
Create a generator function `supplier_generator()` that iterates through all products in the `inventory` and yields each unique supplier name associated with any product.


### Reasoning:
To create the `supplier_generator` function, I will iterate through all categories and their products in the `inventory`. For each product, I will go through its list of suppliers. To ensure that only unique supplier names are yielded, I will maintain a `seen_suppliers` set. Before yielding a supplier name, I will check if it's already in this set; if not, I will add it to the set and then yield it.

**Reasoning**:
Now I will implement the `supplier_generator` function as described, including its definition and a loop to iterate through it, demonstrating its functionality by printing each unique supplier name.



In [18]:
def supplier_generator(data):
    seen_suppliers = set()
    for category_data in data["categories"].values():
        for product in category_data["products"]:
            for supplier in product["suppliers"]:
                if supplier not in seen_suppliers:
                    seen_suppliers.add(supplier)
                    yield supplier

print("\nUnique Supplier Names using supplier_generator:")
for supplier_name in supplier_generator(inventory):
    print(f"- {supplier_name}")


Unique Supplier Names using supplier_generator:
- MobileMasters
- TechCorp
- GadgetHub
- NewSupplierCo
- FashionInc
- StyleWear
- DenimCo


## Summary of Iterator and Generator Functions

This section focused on implementing Python's iteration protocol and generator functions for efficient data traversal.

1.  **`ProductIterator` Class**: This custom iterator class (`ProductIterator`) was defined to implement the iterator protocol (`__iter__` and `__next__`). It allows for iterating through all product names in the `inventory` one by one, providing an iterable object that can be used in `for` loops. Its utility is in providing a standardized way to access elements of a collection without exposing the underlying structure, which can be particularly useful for large datasets where loading all data into memory at once might be inefficient.

2.  **`supplier_generator()` Function**: This generator function was created to iterate through all products and yield each *unique* supplier name. By using a `set` to track `seen_suppliers` and `yield` for returning values, it efficiently produces supplier names on demand, consuming minimal memory. Its utility lies in providing a memory-efficient way to process sequences, especially when dealing with potentially large numbers of suppliers, as it generates values one at a time rather than building an entire list in memory.

### Overall Utility:
These implementations demonstrate fundamental Python concepts for efficient data processing and traversal:
*   **Iterators** (`ProductIterator`) provide a formal, object-oriented way to make objects iterable, useful for structured, stateful iteration.
*   **Generators** (`supplier_generator`) offer a more concise and memory-efficient way to create iterators, particularly suited for large or infinite sequences, as they compute and yield items one by one. Both are crucial for handling large datasets and improving application performance by reducing memory footprint.

## Write a lazy generator stock_stream(min_stock).

### Subtask:
Implement a generator function `stock_stream(min_stock)` that takes a minimum stock level as input. This generator should yield only those product dictionaries whose current stock level is greater than or equal to the `min_stock` threshold.


### Reasoning:
To implement the `stock_stream` generator function, I will iterate through all categories and products in the `inventory` data. For each product, I will check if its 'stock' level meets or exceeds the `min_stock` threshold. If it does, I will `yield` the entire product dictionary. Afterwards, I will demonstrate its usage by creating an instance with a `min_stock` of 20 and print the yielded products.

**Reasoning**:
Now I will implement the `stock_stream` generator function as described, defining it to iterate through the inventory and yield products meeting the minimum stock threshold, and then demonstrate its usage by iterating through the generator with a `min_stock` of 20.



In [19]:
def stock_stream(data, min_stock):
    for category_data in data["categories"].values():
        for product in category_data["products"]:
            if product["stock"] >= min_stock:
                yield product

print(f"\nProducts with stock >= 20:")
# Demonstrate usage with min_stock = 20
for product_info in stock_stream(inventory, 20):
    print(product_info)


Products with stock >= 20:
{'id': 102, 'name': 'Smartphone', 'price': 800.0, 'stock': 28, 'suppliers': ['MobileMasters']}
{'id': 201, 'name': 'T-Shirt', 'price': 25.99, 'stock': 95, 'suppliers': ['FashionInc', 'StyleWear']}
{'id': 202, 'name': 'Jeans', 'price': 49.99, 'stock': 47, 'suppliers': ['DenimCo']}


## Summary of Generator Function `stock_stream`

This section focused on implementing a lazy generator function for efficient data filtering.

1.  **`stock_stream(data, min_stock)` Function**: This generator function takes the inventory data and a minimum stock level as input. It efficiently iterates through all products and yields only those product dictionaries whose current stock level is greater than or equal to the specified `min_stock` threshold. Its utility is in providing a memory-efficient way to filter and retrieve specific product data, especially useful for large inventories where loading all products into memory for filtering might be inefficient. It allows for on-demand processing of products that meet certain stock criteria, which is critical for inventory management tasks like identifying items for reorder or promotional campaigns.

### Overall Utility:
This implementation demonstrates the power of generator functions in Python for creating iterators that produce results on the fly, consuming minimal memory. It's particularly well-suited for scenarios involving large datasets or streams of data where complete lists would be impractical. The `stock_stream` function provides a lean and effective way to monitor inventory levels and react to stock-related conditions.

## Summary of Generator Function `stock_stream`

This section focused on implementing a lazy generator function for efficient data filtering.

1.  **`stock_stream(data, min_stock)` Function**: This generator function takes the inventory data and a minimum stock level as input. It efficiently iterates through all products and yields only those product dictionaries whose current stock level is greater than or equal to the specified `min_stock` threshold. Its utility is in providing a memory-efficient way to filter and retrieve specific product data, especially useful for large inventories where loading all products into memory for filtering might be inefficient. It allows for on-demand processing of products that meet certain stock criteria, which is critical for inventory management tasks like identifying items for reorder or promotional campaigns.

### Overall Utility:
This implementation demonstrates the power of generator functions in Python for creating iterators that produce results on the fly, consuming minimal memory. It's particularly well-suited for scenarios involving large datasets or streams of data where complete lists would be impractical. The `stock_stream` function provides a lean and effective way to monitor inventory levels and react to stock-related conditions.

## Task 3.4: Create a nested generator pipeline.

### Subtask:
Construct a three-step generator pipeline: a generator to get all product dictionaries; a generator to filter products with price > 50.00; and a final generator yielding `(name, price)` tuples.


**Reasoning**:
I need to define the first generator `all_products_generator` that yields all product dictionaries from the `inventory` data, adapting the logic from the previously defined `get_all_products` function.



In [20]:
def all_products_generator(data):
    for category_data in data["categories"].values():
        for product in category_data["products"]:
            yield product

print("Step 1: All products generator created.")

Step 1: All products generator created.


**Reasoning**:
Now I will implement the second generator function `filter_high_price_generator` that takes the output from `all_products_generator` and yields only products with a price greater than 50.00, as specified in the subtask.



In [21]:
def filter_high_price_generator(products_generator):
    for product in products_generator:
        if product['price'] > 50.00:
            yield product

print("Step 2: Filter high price generator created.")

Step 2: Filter high price generator created.


**Reasoning**:
Now I will implement the third and final generator function `product_name_price_generator` that takes the filtered products and yields `(name, price)` tuples, as specified in the subtask.



In [22]:
def product_name_price_generator(filtered_products_generator):
    for product in filtered_products_generator:
        yield (product['name'], product['price'])

print("Step 3: Product name and price generator created.")

Step 3: Product name and price generator created.


**Reasoning**:
I will now construct the full generator pipeline using the three defined generator functions (`all_products_generator`, `filter_high_price_generator`, `product_name_price_generator`) and iterate through the final output to print each `(name, price)` tuple, demonstrating the complete pipeline.



In [23]:
print("\nDemonstrating the generator pipeline (name, price) for products with price > 50.00:")

# Step 1: Get all products
products_gen = all_products_generator(inventory)

# Step 2: Filter high-priced products
filtered_gen = filter_high_price_generator(products_gen)

# Step 3: Extract name and price
final_gen = product_name_price_generator(filtered_gen)

# Iterate and print the results of the pipeline
for name, price in final_gen:
    print(f"- Name: {name}, Price: {price}")


Demonstrating the generator pipeline (name, price) for products with price > 50.00:
- Name: Smartphone, Price: 800.0
- Name: Laptop, Price: 1200.5


## Task 3.5: Demonstrate memory comparison.

### Subtask:
Use `sys.getsizeof()` to compare the memory footprint of creating a list of all product dictionaries using a list comprehension versus creating a generator object using a generator expression for the same data. Display both sizes.


**Reasoning**:
First, I will import the `sys` module, which is necessary for using `sys.getsizeof()` to measure the memory footprint of objects. This prepares the environment for the memory comparison.



In [24]:
import sys

print("sys module imported.")

sys module imported.


**Reasoning**:
I will now create a list of all product dictionaries using a list comprehension and a generator object for all product dictionaries using a generator expression. Then, I will use `sys.getsizeof()` to compare and display their memory footprints to demonstrate the memory efficiency of generators.



In [25]:
all_products_list = [
    product
    for category_data in inventory["categories"].values()
    for product in category_data["products"]
]

all_products_generator_exp = (
    product
    for category_data in inventory["categories"].values()
    for product in category_data["products"]
)

print(f"\nMemory size of all_products_list: {sys.getsizeof(all_products_list)} bytes")
print(f"Memory size of all_products_generator_exp: {sys.getsizeof(all_products_generator_exp)} bytes")
print("Generators typically have a much smaller memory footprint than lists because they generate values on the fly.")


Memory size of all_products_list: 88 bytes
Memory size of all_products_generator_exp: 208 bytes
Generators typically have a much smaller memory footprint than lists because they generate values on the fly.


## Final Task

### Subtask:
Summarize the iterators and generators created, and discuss their utility, especially highlighting the benefits of lazy evaluation and memory efficiency.


## Summary:

### Q&A
The task implicitly asks to summarize the iterators and generators created, discussing their utility, especially highlighting the benefits of lazy evaluation and memory efficiency.

1.  **`ProductIterator`**: This custom iterator class was implemented to iterate through product names one by one. Its utility lies in providing a standardized, object-oriented way to iterate through a collection without loading all data into memory at once, which is beneficial for large datasets.
2.  **`supplier_generator()`**: This generator function was designed to yield unique supplier names efficiently. By using a `set` to track seen suppliers and `yield` for returning values, it processes unique supplier names on demand, consuming minimal memory. This is particularly useful when dealing with a potentially large number of suppliers, as it avoids building a complete list in memory.
3.  **`stock_stream(min_stock)`**: This generator function was created to filter products based on a minimum stock level, yielding only those products that meet the criteria. Its utility is in providing a memory-efficient way to filter and retrieve specific product data, crucial for inventory management tasks without loading the entire inventory into memory for filtering.
4.  **Nested Generator Pipeline**: A three-step pipeline was constructed using generators for product retrieval, filtering by price, and formatting the output. This demonstrates the composability and lazy evaluation benefits of generators, allowing complex data processing without intermediate lists, thus improving memory efficiency for large datasets.
5.  **Memory Comparison**: The comparison between a list comprehension and a generator expression demonstrated that while the initial generator *object* itself might have a comparable or even larger memory footprint than a small list (e.g., 208 bytes for the generator vs. 88 bytes for a small list in the specific test), its true memory efficiency stems from its lazy evaluation, producing values one at a time only when requested, rather than storing all values in memory simultaneously. This is a significant advantage for large or infinite data sequences.

### Data Analysis Key Findings

*   The `ProductIterator` class was successfully implemented, allowing iteration over product names (e.g., "Smartphone", "Laptop", "T-Shirt", "Jeans") from the inventory.
*   The `supplier_generator()` function effectively yielded unique supplier names, such as "MobileMasters", "TechCorp", "GadgetHub", "NewSupplierCo", "FashionInc", "StyleWear", and "DenimCo", demonstrating memory-efficient processing of unique values.
*   The `stock_stream(min_stock)` generator correctly filtered products, yielding only those meeting the minimum stock threshold. For a `min_stock` of 20, it identified "Smartphone" (stock 28), "T-Shirt" (stock 95), and "Jeans" (stock 47).
*   A three-step generator pipeline was successfully constructed to extract names and prices of products with a price greater than \$50.00, yielding "Smartphone" (Price: \$800.0) and "Laptop" (Price: \$1200.5).
*   Memory footprint comparison using `sys.getsizeof()` showed the list comprehension occupying 88 bytes and the generator expression object occupying 208 bytes. This specific result highlights the size of the object itself, while emphasizing that generators' primary memory benefit comes from lazy evaluation over large datasets, not necessarily the initial object size.

### Insights or Next Steps

*   **Embrace Lazy Evaluation for Scalability**: The consistent use of iterators and generators (like `ProductIterator`, `supplier_generator`, `stock_stream`, and the pipeline) effectively demonstrates Python's lazy evaluation capabilities, which are crucial for processing large or streaming datasets by reducing memory footprint and improving performance.
*   **Optimize Memory for Large Datasets**: While `sys.getsizeof()` on a generator object might not always reflect its full memory efficiency for small datasets, understanding that generators avoid materializing entire collections in memory is key. Future work should prioritize generators when dealing with data that could exceed available memory or when only partial data processing is required.


# Task
Implement closures for advanced data handling: create a `make_discount_calculator(rate)` closure that applies a given discount rate to an `original_price`; build a `create_category_filter(cat_name)` closure to filter products by category; use a `lambda` within a closure to format product names as 'Product: <Name>'; develop a nested closure for product ranking based on a `rank_threshold`; and demonstrate value retention in closures with a simple counter. Finally, summarize the utility of these closures and how they encapsulate state.

## Task 6.1: Build a closure `make_discount_calculator(rate)`.

### Subtask:
Create a closure `make_discount_calculator(rate)` that applies a given discount rate to an `original_price`.


### Reasoning:
To create the `make_discount_calculator` closure, I will define an outer function that accepts a `rate` argument. Inside this outer function, I will define a nested function, `calculate_discounted_price`, which will capture the `rate` from its enclosing scope. This nested function will take an `original_price` and compute the discounted price using the captured `rate`. The outer function will then return this nested function. Finally, I will demonstrate the closure by creating two separate discount calculators with different rates and applying them to an example price to verify their functionality.

**Reasoning**:
Now I will implement the `make_discount_calculator` function as described, including the nested function and the demonstration of its usage with different discount rates and an example price.



In [30]:
def make_discount_calculator(rate):
    """Creates a closure that applies a given discount rate to an original price."""
    def calculate_discounted_price(original_price):
        return original_price - (original_price * rate)
    return calculate_discounted_price

# Test the closure
# Create a 10% discount calculator
discount_10_percent = make_discount_calculator(0.10)

# Create a 25% discount calculator
discount_25_percent = make_discount_calculator(0.25)

example_price_1 = 100.00
example_price_2 = 50.00

print(f"\nOriginal price: ${example_price_1:.2f}")
print(f"Price after 10% discount: ${discount_10_percent(example_price_1):.2f}")
print(f"Price after 25% discount: ${discount_25_percent(example_price_1):.2f}")

print(f"\nOriginal price: ${example_price_2:.2f}")
print(f"Price after 10% discount: ${discount_10_percent(example_price_2):.2f}")
print(f"Price after 25% discount: ${discount_25_percent(example_price_2):.2f}")


Original price: $100.00
Price after 10% discount: $90.00
Price after 25% discount: $75.00

Original price: $50.00
Price after 10% discount: $45.00
Price after 25% discount: $37.50


## Task 6.2: Build a closure `create_category_filter(cat_name)`.

### Subtask:
Create a closure `create_category_filter(cat_name)` that takes a category name as input and returns a function. The returned function should take a list of products as input and return only those products belonging to the specified category.

### Reasoning:
To create the `create_category_filter` closure, I will define an outer function that accepts `cat_name`. Inside, I'll define a nested function `filter_products_by_category` that captures `cat_name`. This nested function will then take `all_products` as input and return a filtered list containing only products where `product['category']` matches the captured `cat_name`. The outer function will return this nested function. I will first define a helper function `get_all_products_with_category` to provide a list of products with their category information for the filter function to use. Finally, I will test the closure by creating a filter for 'Electronics' and apply it to an example list of all products.

**Reasoning**:
Before implementing the `create_category_filter` closure, I need a helper function that flattens the nested `inventory` and enriches each product dictionary with its `category` name. This is crucial because the filter function will need this category information to properly filter products.



In [31]:
def get_all_products_with_category(data):
    all_products_categorized = []
    for category_name, category_data in data["categories"].items():
        for product in category_data["products"]:
            # Create a copy of the product and add the category name
            product_with_category = product.copy()
            product_with_category['category'] = category_name
            all_products_categorized.append(product_with_category)
    return all_products_categorized

# Test the helper function
products_with_categories = get_all_products_with_category(inventory)
print("\nAll products with their categories:")
for product in products_with_categories:
    print(f"- {product['name']} (Category: {product['category']})")


All products with their categories:
- Smartphone (Category: Electronics)
- Laptop (Category: Electronics)
- T-Shirt (Category: Clothing)
- Jeans (Category: Clothing)


**Reasoning**:
Now I will implement the `create_category_filter` function as a closure, defining the outer function `create_category_filter` and the nested `filter_products_by_category` function, and then test it by creating a filter for 'Electronics' and applying it to the `products_with_categories` list.



In [32]:
def create_category_filter(cat_name):
    """Creates a closure that filters a list of products by a given category name."""
    def filter_products_by_category(all_products):
        return [product for product in all_products if product.get('category') == cat_name]
    return filter_products_by_category

# Test the closure
# Create a filter for 'Electronics' category
electronics_filter = create_category_filter('Electronics')

# Apply the filter to the list of all products with categories
electronics_products = electronics_filter(products_with_categories)

print(f"\nProducts in 'Electronics' category:")
for product in electronics_products:
    print(f"- {product['name']}")

# Create a filter for 'Clothing' category
clothing_filter = create_category_filter('Clothing')
clothing_products = clothing_filter(products_with_categories)

print(f"\nProducts in 'Clothing' category:")
for product in clothing_products:
    print(f"- {product['name']}")


Products in 'Electronics' category:
- Smartphone
- Laptop

Products in 'Clothing' category:
- T-Shirt
- Jeans


## Task 6.3: Use a `lambda` within a closure to format product names.

### Subtask:
Create a closure that takes a format string as input (e.g., "Product: {}" or "Item: {}"). The returned function should then take a product name and format it according to the captured format string using a `lambda` function.

### Reasoning:
To create a closure that formats product names, I will define an outer function `create_name_formatter` that accepts a `format_string`. Inside, a nested function `format_product_name` will be defined, which captures `format_string` from its enclosing scope. This nested function will then use a `lambda` expression to apply the captured `format_string` to a given `product_name`. The outer function will return this nested function. I will then demonstrate its usage by creating different formatters and applying them to product names.

**Reasoning**:
Now I will implement the `create_name_formatter` closure as described in the subtask and reasoning, defining the outer function and the nested function which uses a `lambda` to format product names. I will then test it by creating multiple formatters and applying them to example product names.



In [33]:
def create_name_formatter(format_string):
    """Creates a closure that formats a product name using a given format string."""
    def format_product_name(product_name):
        # Use a lambda function to apply the captured format_string
        return (lambda name: format_string.format(name))(product_name)
    return format_product_name

# Test the closure
# Create a formatter for 'Product: {}'
product_formatter = create_name_formatter("Product: {}")

# Create a formatter for 'Item (Details): {}'
item_detail_formatter = create_name_formatter("Item (Details): {}")

example_product_1 = "Smartphone"
example_product_2 = "Laptop"

print(f"\nFormatted product name (Product: {{}}): {product_formatter(example_product_1)}")
print(f"Formatted product name (Item (Details): {{}}): {item_detail_formatter(example_product_1)}")

print(f"Formatted product name (Product: {{}}): {product_formatter(example_product_2)}")
print(f"Formatted product name (Item (Details): {{}}): {item_detail_formatter(example_product_2)}")


Formatted product name (Product: {}): Product: Smartphone
Formatted product name (Item (Details): {}): Item (Details): Smartphone
Formatted product name (Product: {}): Product: Laptop
Formatted product name (Item (Details): {}): Item (Details): Laptop


## Task 6.4: Develop a nested closure for product ranking.

### Subtask:
Create a nested closure that helps rank products based on a `rank_threshold`. The outer function `create_rank_evaluator(rank_threshold)` should take a numerical `rank_threshold` (e.g., 500) as input. The first inner function `evaluate_product_rank(product_price)` should take a product's price and return a string indicating if it's 'High Rank' or 'Low Rank' relative to the `rank_threshold`. The second inner function `get_rank_evaluation_function()` should return `evaluate_product_rank`.

### Reasoning:
To create the nested closure for product ranking, I will define an outer function `create_rank_evaluator` that accepts a `rank_threshold`. Inside this, I will define a first inner function `evaluate_product_rank` that takes `product_price` and uses the captured `rank_threshold` to return 'High Rank' or 'Low Rank'. A second inner function `get_rank_evaluation_function` will simply return `evaluate_product_rank`. This structure demonstrates how to encapsulate a specific ranking logic that depends on an initial threshold. I will then test this by creating an evaluator with a threshold and applying it to example product prices.

**Reasoning**:
Now I will implement the nested closure for product ranking as described in the previous reasoning. This will involve defining the outer function, the two nested functions, and then demonstrating its usage with example product prices to evaluate their rank based on a given threshold.



In [34]:
def create_rank_evaluator(rank_threshold):
    """Creates a nested closure to evaluate product rank based on a threshold."""
    def evaluate_product_rank(product_price):
        if product_price >= rank_threshold:
            return 'High Rank'
        else:
            return 'Low Rank'

    def get_rank_evaluation_function():
        return evaluate_product_rank

    return get_rank_evaluation_function

# Test the nested closure
# Create an evaluator with a threshold of 500
rank_evaluator_500 = create_rank_evaluator(500)
get_rank_func_500 = rank_evaluator_500()

# Create an evaluator with a threshold of 1000
rank_evaluator_1000 = create_rank_evaluator(1000)
get_rank_func_1000 = rank_evaluator_1000()

example_price_A = 750.00
example_price_B = 200.00
example_price_C = 1100.00

print(f"\nEvaluating product ranks (Threshold = 500):")
print(f"Price ${example_price_A:.2f}: {get_rank_func_500(example_price_A)}")
print(f"Price ${example_price_B:.2f}: {get_rank_func_500(example_price_B)}")
print(f"Price ${example_price_C:.2f}: {get_rank_func_500(example_price_C)}")

print(f"\nEvaluating product ranks (Threshold = 1000):")
print(f"Price ${example_price_A:.2f}: {get_rank_func_1000(example_price_A)}")
print(f"Price ${example_price_B:.2f}: {get_rank_func_1000(example_price_B)}")
print(f"Price ${example_price_C:.2f}: {get_rank_func_1000(example_price_C)}")


Evaluating product ranks (Threshold = 500):
Price $750.00: High Rank
Price $200.00: Low Rank
Price $1100.00: High Rank

Evaluating product ranks (Threshold = 1000):
Price $750.00: Low Rank
Price $200.00: Low Rank
Price $1100.00: High Rank


## Task 6.5: Demonstrate value retention in closures with a simple counter.

### Subtask:
Create a closure `make_counter()` that returns a function. The returned function, when called, should increment and return a counter that is encapsulated within the closure's scope. This demonstrates how closures retain state between calls.

### Reasoning:
To demonstrate value retention, I will define an outer function `make_counter()` that initializes a `count` variable. Inside `make_counter()`, I will define a nested function `increment_counter()` that increments and returns this `count` variable. The key here is that `increment_counter()` will "remember" the `count` from its enclosing scope, even after `make_counter()` has finished executing. I will then create multiple counter instances to show that each maintains its own independent state.

**Reasoning**:
Now I will implement the `make_counter` closure as described, defining the outer function `make_counter` and the nested function `increment_counter`, and then demonstrate its usage by creating multiple counter instances and observing their independent state.



In [35]:
def make_counter():
    """Creates a closure that encapsulates a counter and returns a function to increment it."""
    count = 0 # This count is encapsulated by the closure
    def increment_counter():
        nonlocal count # Declare count as non-local to modify it
        count += 1
        return count
    return increment_counter

# Test the closure
# Create two independent counter instances
counter1 = make_counter()
counter2 = make_counter()

print("\nDemonstrating independent counters:")

print(f"Counter 1: {counter1()}") # 1
print(f"Counter 1: {counter1()}") # 2
print(f"Counter 2: {counter2()}") # 1
print(f"Counter 1: {counter1()}") # 3
print(f"Counter 2: {counter2()}") # 2


Demonstrating independent counters:
Counter 1: 1
Counter 1: 2
Counter 2: 1
Counter 1: 3
Counter 2: 2


## Summary of Closures and State Encapsulation

This section demonstrated the power and utility of closures in Python for advanced data handling, particularly focusing on their ability to encapsulate state and create specialized functions.

1.  **`make_discount_calculator(rate)`**: This closure creates functions that apply a specific, pre-defined discount rate. Its utility lies in parametrizing a common operation (discount calculation) and generating custom versions of it on the fly. Each returned function 'remembers' the `rate` it was created with, effectively encapsulating that state.

2.  **`create_category_filter(cat_name)`**: This closure generates filter functions for specific product categories. It's useful for creating reusable, category-specific filtering logic without repeatedly passing the category name. The returned function encapsulates the `cat_name`, allowing it to filter product lists based on that remembered category.

3.  **`create_name_formatter(format_string)`**: This closure, utilizing a `lambda` function, generates functions that format product names according to a specified `format_string`. This demonstrates how closures can capture formatting rules, providing flexible and consistent string manipulation. The `format_string` is encapsulated within the returned formatter function.

4.  **`create_rank_evaluator(rank_threshold)`**: This nested closure provides a mechanism to create product ranking functions based on a given `rank_threshold`. The inner `evaluate_product_rank` function encapsulates the `rank_threshold`, allowing it to consistently evaluate product prices against that specific criterion. This shows how closures can build more complex, multi-layered logic while maintaining state.

5.  **`make_counter()`**: This closure explicitly demonstrates state retention. Each call to `make_counter()` creates a new, independent counter function. The `count` variable is encapsulated within each instance of the `increment_counter` function, showing that closures can retain and modify their own internal state across multiple calls, acting like simple objects without explicit class definition.

### Overall Utility and State Encapsulation:

Closures are powerful for:
*   **Encapsulation**: They allow a function to 'remember' variables from its enclosing scope, providing a form of data hiding and state management without the need for full-fledged classes.
*   **Configuration/Parameterization**: They enable the creation of specialized functions from a more general template by fixing certain arguments or settings.
*   **Code Reusability**: By encapsulating logic and state, closures promote cleaner and more modular code, making functions more reusable in different contexts.
*   **Callback Functions**: They are commonly used in event handlers, decorators, and functional programming patterns where a function needs to carry some context with it.

## Summary of Closures and State Encapsulation

This section demonstrated the power and utility of closures in Python for advanced data handling, particularly focusing on their ability to encapsulate state and create specialized functions.

1.  **`make_discount_calculator(rate)`**: This closure creates functions that apply a specific, pre-defined discount rate. Its utility lies in parametrizing a common operation (discount calculation) and generating custom versions of it on the fly. Each returned function 'remembers' the `rate` it was created with, effectively encapsulating that state.

2.  **`create_category_filter(cat_name)`**: This closure generates filter functions for specific product categories. It's useful for creating reusable, category-specific filtering logic without repeatedly passing the category name. The returned function encapsulates the `cat_name`, allowing it to filter product lists based on that remembered category.

3.  **`create_name_formatter(format_string)`**: This closure, utilizing a `lambda` function, generates functions that format product names according to a specified `format_string`. This demonstrates how closures can capture formatting rules, providing flexible and consistent string manipulation. The `format_string` is encapsulated within the returned formatter function.

4.  **`create_rank_evaluator(rank_threshold)`**: This nested closure provides a mechanism to create product ranking functions based on a given `rank_threshold`. The inner `evaluate_product_rank` function encapsulates the `rank_threshold`, allowing it to consistently evaluate product prices against that specific criterion. This shows how closures can build more complex, multi-layered logic while maintaining state.

5.  **`make_counter()`**: This closure explicitly demonstrates state retention. Each call to `make_counter()` creates a new, independent counter function. The `count` variable is encapsulated within each instance of the `increment_counter` function, showing that closures can retain and modify their own internal state across multiple calls, acting like simple objects without explicit class definition.

### Overall Utility and State Encapsulation:

Closures are powerful for:
*   **Encapsulation**: They allow a function to 'remember' variables from its enclosing scope, providing a form of data hiding and state management without the need for full-fledged classes.
*   **Configuration/Parameterization**: They enable the creation of specialized functions from a more general template by fixing certain arguments or settings.
*   **Code Reusability**: By encapsulating logic and state, closures promote cleaner and more modular code, making functions more reusable in different contexts.
*   **Callback Functions**: They are commonly used in event handlers, decorators, and functional programming patterns where a function needs to carry some context with it.

## Final Task

### Subtask:
Summarize the various closures created in this task, highlighting their utility and explaining how they effectively retain and encapsulate state.


## Summary:

### Data Analysis Key Findings

*   **Discount Calculators:** The `make_discount_calculator` closure successfully created specialized functions to apply discounts. For an original price of \$100.00, a 10% discount calculator yielded \$90.00, while a 25% discount calculator resulted in \$75.00, demonstrating effective state encapsulation of the discount rate.
*   **Category Filtering:** The `create_category_filter` closure effectively generated category-specific filters. For instance, filtering products for 'Electronics' successfully returned items like 'Smartphone' and 'Laptop', while 'Clothing' correctly identified 'T-Shirt'. This illustrates the capture of the `cat_name` state.
*   **Dynamic Name Formatting:** The `create_name_formatter` closure, using a lambda function, dynamically formatted product names. With "Product: {}" as the format string, "Smartphone" became "Product: Smartphone", and with "Item (Details): {}", it became "Item (Details): Smartphone", showcasing the encapsulation of the format string.
*   **Product Ranking with Thresholds:** The nested `create_rank_evaluator` closure successfully classified product ranks based on different thresholds. A product priced at \$750.00 was 'High Rank' with a threshold of 500 but 'Low Rank' with a threshold of 1000, confirming the encapsulated `rank_threshold`.
*   **Independent State Retention:** The `make_counter` closure demonstrated robust state retention. Two independent counter instances, `counter1` and `counter2`, maintained separate counts, with `counter1` incrementing to 1, 2, then 3, and `counter2` incrementing to 1, then 2, confirming each closure's ability to hold and modify its own internal state.

### Insights or Next Steps

*   Closures are a versatile tool for creating highly configurable and specialized functions by encapsulating specific parameters or state, significantly enhancing code reusability and modularity without needing to define full classes for simple state management.
*   Further exploration could involve using closures in more complex scenarios, such as creating event handlers with specific contextual data or implementing decorators for function modification, to leverage their state-encapsulation capabilities in advanced programming patterns.
